pkg/providers/launchtemplate/tags.go

```diff
@@ -0,0 +1,65 @@
+/*
+Portions Copyright (c) Microsoft Corporation.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package launchtemplate
+
+import (
+	"strings"
+
+	v1 "sigs.k8s.io/karpenter/pkg/apis/v1"
+
+	"github.com/Azure/karpenter-provider-azure/pkg/apis/v1beta1"
+	"github.com/Azure/karpenter-provider-azure/pkg/operator/options"
+	"github.com/samber/lo"
+)
+
+const (
+	KarpenterManagedTagKey = "karpenter.azure.com_cluster"
+)
+
+var (
+	NodePoolTagKey = strings.ReplaceAll(v1.NodePoolLabelKey, "/", "_")
+)
+
+// TODO: Would like to refactor this out of launchtemplate at some point
+// Tags returns the tags to be applied to a resource (VM, Disk, NIC, etc)
+func Tags(
+	options *options.Options,
+	nodeClass *v1beta1.AKSNodeClass,
+	nodeClaim *v1.NodeClaim,
+) map[string]*string {
+	defaultTags := map[string]string{
+		KarpenterManagedTagKey: options.ClusterName,
+	}
+	// Note: Be careful depending on nodeClaim.Labels here, as we assign some additional labels during the creation
+	// of the static parameters for the launch template. Those labels haven't actually been applied to the nodeClaim yet,
+	// so if you try to use them here you will find they are missing.
+	// For now, we only depend on labels that are added to the nodeClaim itself.
+	if val, ok := nodeClaim.Labels[v1.NodePoolLabelKey]; ok {
+		defaultTags[NodePoolTagKey] = val
```

Comment:
This is a new tag being applied, or did I miss where this is being refactored from?

With this happening at the end of create, it seems a bit circular?
<img width="448" height="81" alt="image" src="https://github.com/user-attachments/assets/8922aead-a035-4201-bf70-47a2ad906de7" />

Reply:
I see it was happening at VM create time in instance:
<img width="528" height="119" alt="image" src="https://github.com/user-attachments/assets/d8e8996c-d63d-4f77-8c17-7f95a3d8ef14" />

Reply:
Yes I refactored this partially because tag application and ordering was all over the place and it was hard to grok. Consolidating to one spot so it can be reused elsewhere (I probably could have broken just this change out into its own PR TBH).